### DISCUSSION TITLE
Cannot install in Homebrew on ARM processor in Intel default prefix -- M2

### DISCUSSION THREAD

**Discussion by mthomp89** [Homebrew/discussions#4397] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.0.6
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7fb3ea78a1e5efb28ad9137ecda5d3b889eb2811
Last commit: 2 years, 1 month ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: e2d726ca1afa11667e76a71fd05c6d0ca85b879d
Core tap last commit: 2 years, 1 month ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.oGm4HWr8Iq/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: dodeca-core 64-bit dunno
Clang: 14.0 build 1403
Git: 2.25.0 => /usr/local/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
XQuartz: 2.7.11 => /opt/X11
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /Users/mk1989f_a/anaconda3/bin/icu-config
  /Users/mk1989f_a/anaconda3/bin/krb5-config
  /Users/mk1989f_a/anaconda3/bin/freetype-config
  /Users/mk1989f_a/anaconda3/bin/xslt-config
  /Users/mk1989f_a/anaconda3/bin/libpng16-config
  /Users/mk1989f_a/anaconda3/bin/libpng-config
  /Users/mk1989f_a/anaconda3/bin/python3.10-config
  /Users/mk1989f_a/anaconda3/bin/xml2-config
  /Users/mk1989f_a/anaconda3/bin/python3-config
  /Users/mk1989f_a/anaconda3/bin/ncursesw6-config
  /Users/mk1989f_a/anaconda3/bin/pcre-config
  /Users/mk1989f_a/anaconda3/bin/nspr-config
  /Users/mk1989f_a/anaconda3/bin/nss-config
  /Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10-config
  /Library/Frameworks/Python.framework/Versions/3.10/bin/python3-config
  /Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9-config
  /Library/Frameworks/Python.framework/Versions/3.9/bin/python3-config
  /Library/Frameworks/Python.framework/Versions/3.8/bin/python3-config
  /Library/Frameworks/Python.framework/Versions/3.8/bin/python3.8-config

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libtcl8.6.dylib
  /usr/local/lib/libtk8.6.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/fakemysql.h
  /usr/local/include/fakepq.h
  /usr/local/include/fakesql.h
  /usr/local/include/itcl.h
  /usr/local/include/itcl2TclOO.h
  /usr/local/include/itclDecls.h
  /usr/local/include/itclInt.h
  /usr/local/include/itclIntDecls.h
  /usr/local/include/itclMigrate2TclCore.h
  /usr/local/include/itclTclIntStubsFcn.h
  /usr/local/include/mysqlStubs.h
  /usr/local/include/odbcStubs.h
  /usr/local/include/pqStubs.h
  /usr/local/include/tcl.h
  /usr/local/include/tclDecls.h
  /usr/local/include/tclOO.h
  /usr/local/include/tclOODecls.h
  /usr/local/include/tclPlatDecls.h
  /usr/local/include/tclThread.h
  /usr/local/include/tclTomMath.h
  /usr/local/include/tclTomMathDecls.h
  /usr/local/include/tdbc.h
  /usr/local/include/tdbcDecls.h
  /usr/local/include/tdbcInt.h
  /usr/local/include/tk.h
  /usr/local/include/tkDecls.h
  /usr/local/include/tkPlatDecls.h

Warning: You are using macOS 13.
We do not provide support for this pre-release version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
pre-release version.


Warning: Your Homebrew's prefix is not /opt/homebrew.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/opt/homebrew).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.


Warning: Ruby version 2.6.10 is unsupported on 13. Homebrew
is developed and tested on Ruby 2.6.3, and may not work correctly
on other Rubies. Patches are accepted as long as they don't cause breakage
on supported Rubies.
```

### Description of issue

I'm new to the Mac family and Brew. Using the new M2 for development, I've attempted to run `brew install git-lfs`. The following error was returned. I've read similar threads for the M1 but could not decipher a clear answer. The `Warning: You are using macOS 13. We do not provide support for this pre-release version.` also provided some confusion.

> Error:
  homebrew-core is a shallow clone.
To `brew update`, first run:
  git -C /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core fetch --unshallow
This command may take a few minutes to run due to the large size of the repository.
This restriction has been made on GitHub's request because updating shallow
clones is an extremely expensive operation due to the tree layout and traffic of
Homebrew/homebrew-core and Homebrew/homebrew-cask. We don't do this for you
automatically to avoid repeatedly performing an expensive unshallow operation in
CI systems (which should instead be fixed to not use shallow clones). Sorry for
the inconvenience!
Error: Cannot install in Homebrew on ARM processor in Intel default prefix (/usr/local)!
Please create a new installation in /opt/homebrew using one of the
"Alternative Installs" from:
  https://docs.brew.sh/Installation
You can migrate your previously installed formula list with:
  brew bundle dump

Please advise. Should I run `brew bundle dump` or perform a complete uninstall/reinstall?

**Comment by gromgit** [comment, marked as answer]
For starters:
>```
>HOMEBREW_VERSION: 3.0.6
>ORIGIN: https://github.com/Homebrew/brew
>HEAD: 7fb3ea78a1e5efb28ad9137ecda5d3b889eb2811
>Last commit: 2 years, 1 month ago
>Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
>Core tap HEAD: e2d726ca1afa11667e76a71fd05c6d0ca85b879d
>Core tap last commit: 2 years, 1 month ago
>Core tap branch: master
>```
your entire Homebrew setup is over two years old (which makes your "new to Brew" comment a bit odd). That's why you got the "macOS 13 is pre-release" warning--it literally didn't exist two years ago! You also missed out on a major usability change: Homebrew has switched to API installs instead of forcing everyone to maintain local core repos, which makes things a whole lot faster while also making your "shallow Git clone" hack a thing of the past.

Your best bet at this point:
1. `arch -x86_64 /usr/local/bin/brew bundle dump --global` to dump your current installed formulae list to `~/.Brewfile`.
2. [Reinstall Homebrew](https://brew.sh/) under the correct prefix (`/opt/homebrew`). Be sure to do the stuff listed at the end of the installation process under **Next Steps**; too many folks forget this and complain about "Homebrew not working" for them.
3. `/opt/homebrew/bin/brew bundle install --global` to replicate your original formulae installation under "new Homebrew". Note that core repos have changed a LOT over the last two years, so some formulae have disappeared or been renamed. Expect errors, and spending some time searching Homebrew for replacements for deprecated, disabled, and just plain _missing_ formulae.

Once you're sure you've gotten your new setup as "whole" as it gets, you can then uninstall the old one:
4. Download the uninstaller script: `curl -fsSL -o /tmp/uninstall.sh https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh`
5. Run it as follows: `bash /tmp/uninstall.sh --path=/usr/local`

**Reply by mthomp89** [reply]
Outstanding! Thank you very much for the detailed explanation and precise commands to use. I was expecting quite a few errors but, surprisingly, none were encountered.

**Reply by Bergruebe** [reply]
Thank you very much, you post helped me a lot after the migration to a Apple silicon Mac via a Time Machine backup. I forgot to run brew bundle on the old Mac without an M chip.

For those who also get the error:
```bash
arch: posix_spawnp: /usr/local/bin/brew: Bad CPU type in executable
```
You need to install Rosetta, you can do it with this command:
```sh
softwareupdate --install-rosetta
```

<hr>

Because I had some difficulties in finding this solution, I added some keywords, so hopefully it is easier to find. 😀

**Reply by wondasom** [reply]
Thank you so much! Just completed the steps you provided and successfully upgraded git, node and typescript without errors 👍🏼

**Reply by dgm** [reply]
I've been using homebrew for over a decade but had no idea that the default install location had changed and the architecture changed so much.    Couldn't the main homepage at least say something about this major upgrade?

**Reply by osalbahr** [reply]
@dgm just to clarify, the install location hasn’t changed in x86 macOS. When ARM macOS rolled out, a decision had to be made to allow both types of executables to co-exist. What do you think a message on the main homepage should say?

**Reply by dgm** [reply]
right, I too migrated to a Apple silicon Mac via Time machine backup.    Existing libraries all worked with the help of rosetta I guess, but when I tried to install a new ruby with rvm, it went down a homebrew rabbit hole installing a ton of things, but it was still installing in /usr/local until I got this error.

**Reply by Kaizer-1** [reply]
Thank you very much, I reinstalled it with your steps and it totally worked.

**Reply by dgm** [reply]
@osalbahr I saw on the install page a small discussion about the new location, but the main page didn't say anything.  I don't know, I get that it could be kind of cluttered and I probably wouldn't even look there until I had a problem.   Is there a way to have brew itself detect this architecture change sooner and point to a page with these instructions?   

So far, after following these instructions, it seems like my homebrew install is more sane, though it also seems that rvm/ruby still has issues finding the correct ssl, fixed by passing some paths: ```rvm install 3.2.2 --with-openssl-dir=$(brew --prefix)/opt/openssl@3```  but then my ruby 3.0.6 was broken, being linked to old files so I had to reinstall linked to openssl 1.1:  ```rvm reinstall 3.0.6 --with-openssl-dir=$(brew --prefix)/opt/openssl@1.1```  as well as doing some gem pristine commands to make sure everything was recompiled with the new architecture.    Makes sense once it's all spelled. out.  :D     (I know, rvm problem isn't homebrew's problem, just leaving it here anyway in case I search for this again in a few years.)

**Reply by osalbahr** [reply]
What would you say the appropriate behavior for your case might look like? if you are interested you may [submit a PR](https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request).

> Is there a way to have brew itself detect this architecture change sooner and point to a page with these instructions?

Yes, `brew` commands can (and do in [`brew.sh`#L22](https://github.com/Homebrew/brew/blob/1d136be229cbb47d1f4e0ff733f11fc9dfa155f5/Library/Homebrew/brew.sh#L22)) detect the architecture.

**Reply by aloifran** [reply]
Thanks a lot! Works perfectly

**Reply by danilomarcus-atm** [reply]
Thanks a lot! saved me!

**Reply by iambudi** [reply]
Thanks a lot! Works nicely.

```bash
The following possible Homebrew files were not deleted:
/usr/local/Frameworks/
/usr/local/Homebrew/
/usr/local/bin/
/usr/local/etc/
/usr/local/go/
/usr/local/include/
/usr/local/lib/
/usr/local/lmm/
/usr/local/opt/
/usr/local/sbin/
/usr/local/share/
/usr/local/var/
You may wish to remove them yourself
``` 
Wondering if it's save to remove them all?

**Reply by osalbahr** [reply]
It is safe to remove them but only after double-checking there aren’t non-Homebrew files there. It won’t break your system but might affect some programs.
